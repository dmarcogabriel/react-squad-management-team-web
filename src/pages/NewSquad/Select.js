import React, { useState, useEffect } from 'react';
import { MdKeyboardArrowUp, MdKeyboardArrowDown } from 'react-icons/md';

const ICON_SIZE = 24;

const Select = ({ value, options = [], id, onChange = () => {} }) => {
  const [showOptions, setShowOptions] = useState(false);
  const [defaultValue, setDefaultValue] = useState('Chose one...');
  const [optionsList, setOptionsList] = useState(options);

  const handleClickOutside = (e) => {
    const { classList } = e.target;

    if (
      !classList.contains('option') &&
      !classList.contains('selected-option')
    ) {
      setShowOptions(false);
    }
  };

  const handleSelectOption = (e) => {
    setOptionsList(
      optionsList.map((option) => {
        return option.id === e.id
          ? { ...option, selected: true }
          : { ...option, selected: false };
      })
    );
    setDefaultValue(e.label);
    setShowOptions(false);
    onChange(e);
  };

  const isSelected = (selected) =>
    selected ? 'bg-primary-dark text-white' : 'bg-white text-black';

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);

    return () => document.removeEventListener('mousedown', handleClickOutside);
  });

  useEffect(() => {
    if (value) {
      setDefaultValue(value);
    }
  }, [value]);

  return (
    <div className="relative w-40">
      <button
        type="button"
        className="bg-white border-2 border-gray-600 rounded cursor-pointer"
        onClick={() => setShowOptions(!showOptions)}
      >
        <div className="flex">
          <p className="py-1 w-24">{defaultValue}</p>

          <div className="border-l-2 border-gray-600 flex items-center justify-center px-1">
            {showOptions ? (
              <MdKeyboardArrowUp className="text-black" size={ICON_SIZE} />
            ) : (
              <MdKeyboardArrowDown className="text-black" size={ICON_SIZE} />
            )}
          </div>
        </div>
      </button>

      {showOptions && (
        <ul id={id} className="absolute bg-white shadow-xl z-50">
          {optionsList.map((option) => (
            <li className="w-40" key={option.id}>
              <button
                type="button"
                className={`option flex justify-center w-full py-3 font-base cursor-pointer ${isSelected(
                  option.selected
                )}`}
                data-name={option.label}
                key={option.id}
                onClick={() => handleSelectOption(option)}
              >
                {option.label}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Select;
