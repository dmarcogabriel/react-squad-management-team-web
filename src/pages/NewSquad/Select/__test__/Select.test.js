import { render } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import userEvent from '@testing-library/user-event';
import Select from '..';

const mockOptions = [
  {
    id: 'm1',
    label: 'Mock Label',
    selected: false,
  },
  {
    id: 'm2',
    label: 'Mock Label',
    selected: true,
  },
];

describe('<Select />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Select />, div);
  });

  it('renders value correctly', () => {
    const { getByTestId } = render(<Select value="Test Value" />);

    const value = getByTestId('value');

    expect(value).toHaveTextContent('Test Value');
  });

  it('renders options correctly', () => {
    const { getByTestId } = render(<Select options={mockOptions} />);

    const button = getByTestId('button');
    userEvent.click(button);

    expect(getByTestId('option-m1')).toBeInTheDocument();
  });

  it('executes option button click', () => {
    let result;
    const onChange = (e) => {
      result = e;
    };

    const { getByTestId } = render(
      <Select onChange={onChange} options={mockOptions} />
    );

    const button = getByTestId('button');
    userEvent.click(button);

    const optionButton = getByTestId('option-button-m1');
    userEvent.click(optionButton);

    expect(result).toEqual(mockOptions[0]);
  });

  it('renders selected option classes correctly', () => {
    const { getByTestId } = render(<Select options={mockOptions} />);

    const button = getByTestId('button');
    userEvent.click(button);

    expect(getByTestId('option-button-m2').className).toEqual(
      'option flex justify-center w-full py-3 font-base cursor-pointer bg-primary-dark text-white'
    );
  });

  it('renders not selected option classes correctly', () => {
    const { getByTestId } = render(<Select options={mockOptions} />);

    const button = getByTestId('button');
    userEvent.click(button);

    expect(getByTestId('option-button-m1').className).toEqual(
      'option flex justify-center w-full py-3 font-base cursor-pointer bg-white text-black'
    );
  });

  it('renders arrow icon correctly', () => {
    const { getByTestId } = render(<Select options={mockOptions} />);

    const button = getByTestId('button');
    userEvent.click(button);

    expect(getByTestId('arrow-icon')).toBeInTheDocument();
  });

  it('match snapshot', () => {
    expect(shallow(<Select options={mockOptions} />)).toMatchSnapshot();
  });
});
