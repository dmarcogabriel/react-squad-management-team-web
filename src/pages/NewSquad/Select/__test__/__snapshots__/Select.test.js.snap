// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> match snapshot 1`] = `
<div
  className="relative w-40"
>
  <button
    className="bg-white border-2 border-gray-600 rounded cursor-pointer"
    data-testid="button"
    onClick={[Function]}
    type="button"
  >
    <div
      className="flex"
    >
      <p
        className="py-1 w-24"
        data-testid="value"
      >
        Chose one...
      </p>
      <div
        className="border-l-2 border-gray-600 flex items-center justify-center px-1"
      >
        <svg
          className="text-black"
          fill="currentColor"
          height={24}
          size={24}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
        </svg>
      </div>
    </div>
  </button>
</div>
`;
